File: community_compute/desktop/engine.py
# -*- coding: utf-8 -*-
"""The worker engine — a QThread running the resilient pull-loop (LINE model).

v1.0.2 rewrite. The previous build called the control plane with the OLD
job-batch signatures while client.py had already moved to single lines:

    client.submit(worker, item["job_id"], item["out"], proxy)   # 4 args
    def submit(worker_id, out, proxy)                           # 3 params

so the job id was passed AS the payload, the server answered 400, and the engine
treated that as "rejected → drop it locally". Every finished translation was
silently discarded, and the heartbeat was never sent at all, so the server kept
reclaiming this device's lines while it was alive. Both are fixed here.

One tick:
    1. heartbeat if the server's live interval has elapsed  (ONE cheap write)
    2. flush the outbox   (submit finished lines; a blip → stay buffered)
    3. refill the inbox   (claim ahead so a blip never stalls work; capped)
    4. translate a slice from the inbox with the volunteer's own providers
Off = idle, but finished work is still pushed. A drop to the control plane does
NOT stop translation — the worker keeps going from its local buffer.
"""
from __future__ import annotations

import time

from PySide6.QtCore import QThread, Signal

import client
import keystore
import providers
from config import POLL_IDLE_S, POLL_MAX_S, PREFETCH_LINES, TRANSLATE_SLICE
from state import State


class Engine(QThread):
    status = Signal(dict)

    def __init__(self, state: State):
        super().__init__()
        self._st = state
        self._alive = True
        self._on = bool(state.settings().get("enabled"))
        self._online = False
        self._providers_ok = False
        self._note = ""
        self._stage = 0            # 0 pull · 1 translate · 2 check · 3 send
        self._busy = False
        self._last_beat = 0.0
        self._enrolled = False
        self._started_at = 0.0
        self._session_lines = 0

    # -------- control from the UI thread
    def set_on(self, on: bool) -> None:
        self._on = on
        self._st.set_setting("enabled", on)
        if on:
            self._started_at = time.time()
        else:
            self._started_at = 0.0
            self._busy = False
        self._emit()

    def stop(self) -> None:
        self._alive = False

    def uptime(self) -> int:
        return int(time.time() - self._started_at) if self._started_at else 0

    # -------- status
    def _emit(self) -> None:
        keys = keystore.load()
        snap = self._st.snapshot()
        self.status.emit({
            "on": self._on,
            "online": self._online,
            "providers_ok": self._providers_ok,
            "has_keys": bool(providers.available(keys)),
            "n_keys": len(providers.available(keys)),
            "inbox": snap["inbox"],
            "outbox": snap["outbox"],
            "lines": snap["lines_done"],
            "session": self._session_lines,
            "by_provider": snap["by_provider"],
            "note": self._note,
            "stage": self._stage,
            "busy": self._busy,
            "uptime": self.uptime(),
            "blocked": client.BLOCKED,
            "config": dict(client.SERVER_CONFIG),
        })

    def _set_stage(self, stage: int, busy: bool = True) -> None:
        self._stage, self._busy = stage, busy
        self._emit()

    # -------- the loop
    def run(self) -> None:
        backoff = POLL_IDLE_S
        if self._on:
            self._started_at = time.time()
        self._emit()

        while self._alive:
            keys = keystore.load()
            proxy = self._st.settings().get("proxy", "") or ""
            provs = providers.available(keys)

            if not self._on:
                self._flush_outbox(proxy)          # still push finished work
                self._note = "כבוי"
                self._set_stage(0, False)
                self._sleep(POLL_IDLE_S)
                backoff = POLL_IDLE_S
                continue

            if not provs:
                self._note = "צריך להוסיף מפתח לפחות מספק אחד"
                self._set_stage(0, False)
                self._sleep(POLL_IDLE_S)
                continue

            if client.BLOCKED:
                self._note = "המכשיר נחסם על ידי המנהל"
                self._set_stage(0, False)
                self._sleep(POLL_MAX_S)
                continue

            # 0. enroll once (and again whenever the server asks us to)
            if not self._enrolled or client.NEEDS_REENROLL:
                try:
                    client.enroll(self._st.worker_id, "windows", proxy)
                    self._enrolled, self._online = True, True
                    self._last_beat = time.time()
                except client.NetworkError:
                    self._online = False
                    self._note = "אין קשר לשרת — מנסה שוב"
                    self._set_stage(0, False)
                    self._sleep(min(backoff, POLL_MAX_S))
                    backoff = min(backoff * 2, POLL_MAX_S)
                    continue
                except client.ApiError as e:
                    self._note = "השרת דחה: " + str(e)[:60]
                    self._set_stage(0, False)
                    self._sleep(min(backoff, POLL_MAX_S))
                    backoff = min(backoff * 2, POLL_MAX_S)
                    continue

            # 1. heartbeat on the server's OWN live interval (retunable, no rebuild)
            beat = max(60, int(client.SERVER_CONFIG.get("heartbeat_seconds") or 300))
            if self._online and time.time() - self._last_beat >= beat:
                try:
                    if client.renew(self._st.worker_id, proxy):
                        self._last_beat = time.time()
                    elif client.NEEDS_REENROLL:
                        self._enrolled = False      # the server forgot us → re-enroll next tick
                except client.NetworkError:
                    self._online = False
                except client.ApiError:
                    pass

            did_work = False

            # 2. push whatever is already translated
            if self._st.outbox_count():
                self._set_stage(3)
            self._flush_outbox(proxy)

            # 3. refill the buffer (claim ahead) while online and under the cap.
            #    The size is the SERVER's call (it ignores our `max` on purpose), and it
            #    already bounds us by batch_size + max_inflight — so keep EVERYTHING it
            #    hands over. Slicing the reply here would strand the discarded lines:
            #    they are already leased to this worker, so nobody else can take them
            #    until the lease expires, and we would never translate them.
            if self._online and self._st.inbox_count() < PREFETCH_LINES:
                self._set_stage(0)
                try:
                    batch = int(client.SERVER_CONFIG.get("batch_size") or 50)
                    lines = client.claim(self._st.worker_id, batch, proxy)
                    if lines:
                        self._st.add_inbox(lines)
                        self._online = True
                except client.NetworkError:
                    self._online = False
                except client.ApiError as e:
                    self._note = "השרת דחה: " + str(e)[:60]

            # 4. translate a slice locally (needs the PROVIDERS, not the server)
            slice_ = self._st.take_inbox(TRANSLATE_SLICE)
            if slice_:
                self._set_stage(1)
                sysmsg = slice_[0].get("sys") or ""
                items = {ln["id"]: ln.get("src") or "" for ln in slice_}
                try:
                    out, counts = providers.translate_batch(keys, sysmsg, items, proxy)
                except Exception:
                    out, counts = {}, {}
                if out:
                    self._set_stage(2)
                    self._st.add_outbox(out, counts)
                    self._providers_ok = True
                    did_work = True
                    missing = [ln for ln in slice_ if ln["id"] not in out]
                    if missing:
                        self._st.put_back(missing)   # a dropped line is retried, never lost
                    self._set_stage(3)
                    self._flush_outbox(proxy)
                else:
                    self._st.put_back(slice_)
                    self._providers_ok = False

            # 5. status + pacing
            if did_work:
                backoff = POLL_IDLE_S
                self._note = "פעיל" if self._online else "פעיל — נאגר מקומית (השרת לא זמין)"
                self._emit()
            else:
                if not self._online and self._st.inbox_count() == 0:
                    self._note = "אין קשר לשרת — ממתין וחוזר אוטומטית"
                elif not self._providers_ok and self._st.inbox_count() > 0:
                    self._note = "אין קשר לספקים — ממתין"
                else:
                    self._note = "אין כרגע עבודה בתור"
                backoff = min(backoff * 2, POLL_MAX_S)
                self._set_stage(0, False)
                self._sleep(backoff)

        # graceful exit: push what we have, then hand the rest back to the pool
        proxy = self._st.settings().get("proxy", "") or ""
        self._flush_outbox(proxy)
        try:
            if self._enrolled:
                client.release(self._st.worker_id, proxy)
        except Exception:
            pass

    # -------- helpers
    def _flush_outbox(self, proxy: str) -> None:
        pending = self._st.peek_outbox()
        if not pending:
            return
        try:
            accepted = client.submit(self._st.worker_id, pending, proxy)
            # Drop ALL submitted ids: an id the server refused is one we no longer
            # hold (its lease moved on), so retrying it forever would wedge the
            # outbox. Only the ACCEPTED ones are credited.
            self._st.drop_outbox(list(pending), accepted)
            self._session_lines += accepted
            self._online = True
        except client.NetworkError:
            self._online = False        # server down → keep everything buffered
        except client.ApiError:
            self._st.drop_outbox(list(pending), 0)

    def _sleep(self, secs: float) -> None:
        end = time.time() + secs
        while self._alive and time.time() < end:
            time.sleep(0.2)

File: community_compute/android/lib/engine.dart
// The worker engine — a resilient LINE-MODEL pull-loop as a ChangeNotifier.
// Mirrors desktop/engine.py. Each tick: flush outbox -> heartbeat -> refill the
// line buffer -> translate one batch -> flush. A drop to the control plane does
// NOT stop translation; work buffers locally and syncs on reconnect. The store
// (Supabase) owns what is out and what is in: a claimed line that never comes
// back lease-expires and returns to the pool for OTHER workers - no reslice.
//
// Exposes a live PHASE so the UI can show, on a rainbow arc, exactly what is
// happening now: fetch -> translate/review (+auto-check) -> send.
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'client.dart' as api;
import 'config.dart';
import 'fg_service.dart';
import 'keystore.dart';
import 'providers.dart' as prov;
import 'state.dart';
import 'theme.dart' show Prefs;

class Engine extends ChangeNotifier {
  final AppState st;
  bool _alive = true;
  bool on;
  bool online = false;
  bool providersOk = false;
  int nKeys = 0;

  // live pipeline phase (drives the arc + notification)
  String phase = 'off'; // off|nokeys|fetch|translate|review|check|submit|idle|buffer|provfail
  String mode = 'translate'; // translate | review (from the job's sys)
  String provDetail = '';
  int _provFails = 0;
  DateTime _lastRenew = DateTime.fromMillisecondsSinceEpoch(0);
  DateTime _lastNotif = DateTime.fromMillisecondsSinceEpoch(0);
  DateTime _lastIpAt = DateTime.fromMillisecondsSinceEpoch(0);   // live IP refresh
  int _emptyClaims = 0;   // consecutive claims that came back empty (real drain vs a blip)
  bool _releasedOnOff = false;

  Timer? _uiTimer;   // refreshes the live data + rotates the explanations
  int _variant = 0;  // which phrasing of the current phase is shown
  int _tick = 0;     // 3s ticks; the explanation only changes every ~10s

  Engine(this.st) : on = st.enabled;

  Future<void> setOn(bool v) async {
    on = v;
    _releasedOnOff = false;
    phase = v ? 'idle' : 'off';
    await st.setEnabled(v);
    _pushNotif(force: true);
    notifyListeners();
  }

  void stop() { _alive = false; _uiTimer?.cancel(); }

  Future<void> refreshKeys() async {
    nKeys = KeyStore.available(await KeyStore.load()).length;
    notifyListeners();
  }

  void _setPhase(String p) { phase = p; notifyListeners(); _pushNotif(); }

  // ---- what the ring reads (4 stages: fetch · translate/review · check · send) --
  int get ringStage => switch (phase) {
        'fetch' => 0,
        'translate' || 'review' => 1,
        'check' => 2,
        'submit' => 3,
        _ => -1,
      };
  int get arcSegment => switch (phase) {
        'fetch' => 0,
        'translate' || 'review' || 'check' => 1,
        'submit' => 2,
        _ => -1,
      };
  // true only when the SERVER really had nothing for us (≥2 empty claims) — so the
  // UI never falsely says "no lines in queue" on a momentary blip.
  bool get queueEmpty => _emptyClaims >= 2;
  bool get arcRunning => on && (phase == 'fetch' || phase == 'translate' ||
      phase == 'review' || phase == 'check' || phase == 'submit');

  String get stageTitle => switch (phase) {
        'off' => 'כבוי',
        'nokeys' => 'צריך מפתח',
        'fetch' => 'שולף שורות',
        'translate' => 'מתרגם',
        'review' => 'בודק ומשפר',
        'check' => 'בדיקת תקינות',
        'submit' => 'שולח בחזרה',
        'buffer' => 'נאגר מקומית',
        'provfail' => 'תקלת ספק',
        _ => queueEmpty ? 'אין שורות בתור' : 'מתחבר לתור',
      };
  // rotating, plural/impersonal explanations — a fresh phrasing every few seconds
  // so the screen always feels alive, never stuck.
  List<String> _subVariants() => switch (phase) {
        'off' => const ['הפעילו את המתג כדי לתרום כוח-תרגום', 'המתג כבוי - שום עבודה לא רצה כרגע'],
        'nokeys' => const ['הוסיפו מפתח של ספק אחד לפחות בעמוד «מפתחות»',
              'בלי מפתח אין במה לתרגם - הוסיפו אחד בעמוד «מפתחות»'],
        'fetch' => const ['מושכים מנת שורות חדשה מהתור', 'התור שולח את המנה הבאה לתרגום',
              'מקבלים שורות טריות מהשרת'],
        'translate' => const ['המפתחות שלכם מתרגמים לעברית מול כמה שפות',
              'משווים לכמה שפות ובוחרים את הניסוח הטוב ביותר', 'מייצרים תרגום עברי איכותי'],
        'review' => const ['סוקרים תרגום קיים ומתקנים רק טעויות אמת',
              'עוברים על השורות ומשפרים מה שצריך', 'בקרת איכות - תרגום טוב נשאר כמו שהוא'],
        'check' => const ['מוודאים שהתוצאות שלמות ושהטוקנים נשמרו',
              'בדיקת תקינות אוטומטית לפני השליחה', 'מאמתים כל שורה לפני שהיא נשלחת'],
        'submit' => const ['מחזירים את השורות המתורגמות אל השרת',
              'שולחים את המנה שהושלמה בחזרה', 'המנה מסתנכרנת חזרה לתור'],
        'buffer' => const ['אין קשר לשרת - ממשיכים ואוגרים, יסתנכרן אוטומטית',
              'עובדים במצב לא-מקוון - הכול יישלח כשהחיבור יחזור',
              'שומרים את התוצאות מקומית עד שהשרת יחזור'],
        'provfail' => [provDetail.isEmpty ? 'אין קשר לספקים - מנסים שוב' : provDetail],
        _ => queueEmpty
            ? const ['אין כרגע שורות פנויות בתור - יתחדש כשיתווספו שורות',
                  'התור התרוקן לרגע - מוכנים לתרגם ברגע שיתווספו שורות']
            : const ['מתחברים לתור ובודקים שורות חדשות', 'מתארגנים לסבב הבא - עוד רגע ממשיכים'],
      };
  String get stageSub { final v = _subVariants(); return v[_variant % v.length]; }

  Future<void> start() async {
    var backoff = Config.pollIdleS;
    // refresh the on-screen data every 3s (rate / buffer / uptime), but only
    // change the wording of the explanation every ~10s so it never looks jittery.
    _uiTimer ??= Timer.periodic(const Duration(seconds: 3), (_) {
      _tick++;
      if (_tick % 3 == 0) _variant++;   // new phrasing ~every 9-10s
      notifyListeners();
      _pushNotif();
    });
    notifyListeners();
    while (_alive) {
      try {
      final keys = await KeyStore.load();
      nKeys = KeyStore.available(keys).length;
      final proxy = st.proxy;

      if (!on) {
        await _flush(proxy);
        if (!_releasedOnOff && st.inbox.isNotEmpty) {
          try {
            await api.release(st.workerId, proxy);
            await st.clearInbox();
            _releasedOnOff = true; online = true;
          } on api.NetworkError { online = false; } on api.ApiError { _releasedOnOff = true; }
        }
        _setPhase('off');
        await _sleep(Config.pollIdleS);
        backoff = Config.pollIdleS;
        continue;
      }
      if (nKeys == 0) {
        _setPhase('nokeys');
        await _sleep(Config.pollIdleS);
        continue;
      }

      // 0. enroll
      try {
        await api.enroll(st.workerId, proxy);
        online = true;
        // refresh the public IP live — on first run and whenever it may have
        // changed (network switch); updates the display only if it actually differs.
        if (st.ip.isEmpty || DateTime.now().difference(_lastIpAt).inSeconds > 90) {
          _lastIpAt = DateTime.now();
          unawaited(_refreshIp(proxy));
        }
      } on api.NetworkError {
        online = false;
      } on api.ApiError catch (e) {
        provDetail = 'השרת דחה: ${e.m}'; _setPhase('provfail');
        await _sleep(backoff);
        backoff = (backoff * 2).clamp(Config.pollIdleS, Config.pollMaxS);
        continue;
      }

      var didWork = false;

      // 1. flush finished results
      if (st.outbox.isNotEmpty) { _setPhase('submit'); }
      await _flush(proxy);

      // 2. heartbeat — ONE cheap write, at the interval the SERVER dictates
      // (ServerConfig.heartbeatS, default 5 min). Only while we can actually
      // translate, so we never hold lines hostage.
      if (online && providersOk && st.inbox.isNotEmpty &&
          DateTime.now().difference(_lastRenew).inSeconds >= ServerConfig.heartbeatS) {
        try {
          await api.renew(st.workerId, proxy); _lastRenew = DateTime.now(); online = true;
          if (api.needsReenroll) {                       // our worker row vanished
            await api.enroll(st.workerId, proxy);
          }
        }
        on api.NetworkError { online = false; } on api.ApiError { /* ignore */ }
      }

      // 3. refill. Small PROBE while providers are failing so we self-recover.
      if (online && st.inbox.length < Config.prefetchLines) {
        if (st.inbox.isEmpty) { _setPhase('fetch'); }
        final n = _provFails >= 2 ? 5 : ServerConfig.batchSize;
        try {
          final jobs = await api.claim(st.workerId, n, proxy);
          if (jobs.isNotEmpty) { await st.addInbox(jobs); online = true; _emptyClaims = 0; }
          else if (st.inbox.isEmpty) { _emptyClaims++; }   // server had nothing for us
        } on api.NetworkError { online = false; }
        on api.ApiError catch (e) { provDetail = 'השרת דחה: ${e.m}'; }
      }

      // 4. translate ONE batch (providers, not the server)
      final batch = await st.takeBatch(Config.workBatch);
      if (batch.isNotEmpty) {
        mode = batch.first.sys.contains('REVIEW') ? 'review' : 'translate';
        _setPhase(mode == 'review' ? 'review' : 'translate');
        final items = {for (final j in batch) j.id: j.src};
        Map<String, String> out = {};
        Map<String, int> counts = {};
        Map<String, String> statuses = {};
        try {
          final r = await prov.translateBatch(keys, batch.first.sys, items, proxy);
          out = r.$1; counts = r.$2; statuses = r.$3;
        } catch (_) {}
        if (out.isNotEmpty) {
          _setPhase('check');
          await st.addOutbox(out, counts);
          providersOk = true; didWork = true; _provFails = 0; provDetail = '';
          final missed = batch.where((j) => !out.containsKey(j.id)).toList();
          if (missed.isNotEmpty) await st.requeueFront(missed);
          _setPhase('submit');
          await _flush(proxy);
        } else {
          providersOk = false; _provFails++;
          provDetail = _statusNote(statuses);
          await st.requeueFront(batch);
          // a SINGLE momentary blip (rate-limit / network hiccup) recovers on the
          // next retry — don't flash a red "provider fault"; only surface it once
          // it has failed ≥2 times in a row (a genuine sustained problem).
          _setPhase(_provFails >= 2 ? 'provfail' : (online ? 'idle' : 'buffer'));
          if (online && _provFails >= 2) {
            try { await api.release(st.workerId, proxy); await st.clearInbox(); }
            on api.NetworkError { online = false; } on api.ApiError { /* ignore */ }
          }
        }
      }

      if (didWork) {
        backoff = Config.pollIdleS;
        _setPhase(online ? 'idle' : 'buffer');   // brief resting label between batches
      } else {
        if (!providersOk && _provFails >= 2 && provDetail.isNotEmpty) { _setPhase('provfail'); }
        else if (!online) { _setPhase('buffer'); }
        else { _setPhase('idle'); }
        // when we're online and the queue is NOT confirmed-drained, re-check SOON —
        // a spurious empty response / blip must never strand us on a long idle sleep
        // while thousands of lines wait. Only back off far when offline / truly drained.
        final ceiling = (online && !queueEmpty) ? 18 : Config.pollMaxS;
        backoff = (backoff * 2).clamp(Config.pollIdleS, ceiling);
      }
      if (!didWork) await _sleep(backoff);
      } catch (_) {
        // NEVER let an unexpected error kill the pull-loop (a dead loop = a frozen
        // "idle" screen forever). Breathe and continue — the queue keeps its lines.
        online = false;
        await _sleep(Config.pollIdleS);
      }
    }
    await _flush(st.proxy);
  }

  String _statusNote(Map<String, String> s) {
    const label = {'groq': 'Groq', 'sambanova': 'SambaNova', 'nim': 'NIM'};
    final bad = <String>[];
    s.forEach((p, code) {
      if (code == 'ok') return;
      final why = switch (code) {
        '401' || '403' => 'מפתח שגוי/נדחה',
        '429' => 'עומס - מנסה שוב',
        '402' => 'דורש תשלום',
        '503' => 'שירות לא זמין',
        'http' => 'שגיאת שרת',
        'parse' => 'תשובה לא תקינה',
        'timeout' => 'איטי מדי (זמן קצוב)',
        'cooldown' => 'בהמתנה',
        _ => 'שגיאת רשת',
      };
      bad.add('${label[p] ?? p}: $why');
    });
    return bad.isEmpty ? 'אין קשר לספקים - מנסה שוב' : bad.join(' · ');
  }

  Future<void> _flush(String proxy) async {
    if (st.outbox.isEmpty) return;
    final snap = st.outboxSnapshot();
    try {
      final committed = await api.submit(st.workerId, snap, proxy);
      await st.clearOutbox(snap.keys, committed);
      online = true;
    } on api.NetworkError {
      online = false;
    } on api.ApiError {
      await st.clearOutbox(snap.keys, 0);
    }
  }

  Future<void> _refreshIp(String proxy) async {
    try {
      final ip = await api.fetchIp(proxy);
      if (ip.isNotEmpty && ip != st.ip) { await st.setIp(ip); notifyListeners(); }
    } catch (_) {}
  }

  /// Push the live counters into the persistent notification (throttled).
  void _pushNotif({bool force = false}) {
    if (!Prefs.I.notif) return;
    final now = DateTime.now();
    if (!force && now.difference(_lastNotif).inSeconds < 12) return;
    _lastNotif = now;
    if (!on) { FgService.updateStats(paused: true); return; }
    FgService.updateStats(
      lastHour: st.linesLastHour(),
      ratePerMin: st.ratePerMin(),
      stage: stageTitle,
      total: st.linesDone,
      uptime: st.uptimeText(),
    );
  }

  Future<void> _sleep(int secs) async {
    final end = DateTime.now().add(Duration(seconds: secs));
    while (_alive && DateTime.now().isBefore(end)) {
      await Future.delayed(const Duration(milliseconds: 250));
    }
  }
}
